runtime/android/core/src/org/xwalk/core/XWalkActivity.java

```diff
@@ -5,16 +5,720 @@
 package org.xwalk.core;
 
 import android.app.Activity;
+import android.app.AlertDialog;
+import android.app.Dialog;
+import android.app.DownloadManager;
+import android.app.DownloadManager.Request;
+import android.app.DownloadManager.Query;
+import android.app.ProgressDialog;
+import android.content.ActivityNotFoundException;
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.DialogInterface;
+import android.content.Intent;
+import android.content.IntentFilter;
 import android.content.res.Resources;
+import android.database.Cursor;
+import android.net.Uri;
+import android.os.AsyncTask;
+import android.os.Bundle;
+import android.os.Environment;
+import android.util.Log;
+
+import java.io.File;
+import java.lang.Thread;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.LinkedList;
+import java.util.Iterator;
+
+import org.xwalk.core.XWalkCoreWrapper.XWalkCoreStatus;
 
 /**
  * XWalkActiviy is to support cross package resource loading.
  * It provides method to allow overriding getResources() behavior.
  */
-public abstract class XWalkActivity extends Activity {
+public abstract class XWalkActivity extends Activity
+        implements XWalkCoreWrapper.XWalkCoreListener {
+    private static final String XWALK_CORE_APK = "XWalkRuntimeLib.apk";
+    
+    private static final String XWALK_CORE_MARKET_URL =
+            "market://details?id=" + XWalkCoreWrapper.XWALK_CORE_PACKAGE;
+
+    private static final String XWALK_CORE_DOWNLOAD_URL =
+            "http://10.0.2.2/" + XWALK_CORE_APK; // localhost for test
+
+    private DownloadTask mDownloadTask;
+    private Dialog mActiveDialog;
+    private boolean mIsDisplayed;
+
+    private XWalkCoreWrapper mCoreWrapper;
+    private LinkedList<Object> mReservedObjects;
+    private LinkedList<ReflectMethod> mReservedMethods;
+
+    public enum XWalkLibraryStatus {
+        MATCHED,
+        NOT_FOUND,
+        NEWER_VERSION,
+        OLDER_VERSION,
+        COMPRESSED,
```

Comment:
The XWalkLibraryStatus filed means the result of jar version check. Jar version and whether native so is compressed are two different things. It's better to treat them differently